File: lib/hibike_message.cpp
#include "hibike_message.h"

//// CLASS METHOD IMPLEMENTATIONS //////////////////////////////////////////////

//// FUN WITH CHECKSUMS //////////////////////////////////////////////
void SubscriptionRequest::calculateChecksum() {
  uint8_t cs = 0;
  cs ^= messageId;
  cs ^= controllerId;
  cs ^= subscriptionDelay & 0xFF;
  cs ^= (subscriptionDelay >> 8) & 0xFF;
  cs ^= (subscriptionDelay >> 16) & 0xFF;
  cs ^= (subscriptionDelay >> 24) & 0xFF;
  checksum = cs;
  checksumCalculated = true;
}

void SubscriptionResponse::calculateChecksum() {
  uint8_t cs = 0;
  cs ^= messageId;
  cs ^= controllerId;
  checksum = cs;
  checksumCalculated = true;
}

void SubscriptionSensorUpdate::calculateChecksum() {
  uint8_t cs = 0;
  cs ^= messageId;
  cs ^= controllerId;
  cs ^= sensorTypeId;
  cs ^= sensorReadingLength & 0xFF
  cs ^= (sensorReadingLength >> 8) & 0xFF
  for (int i = 0; i < sensorReadingLength; i++) {
    cs ^= *(dataPtr+i) & 0xFF;
  }
  checksum = cs;
  checksumCalculated = true;
}

void Error::calculateChecksum() {
  uint8_t cs = 0;
  cs ^= messageId;
  cs ^= controllerId;
  cs ^= errorCode
  checksum = cs;
  checksumCalculated = true;
}

//// MESSAGE SENDING /////////////////////////////////////////////////

// Note: we assume that Serial (from the Arduino libraries) has already
// been imported
void SubscriptionRequest::send() {
  Serial.write(messageId);
  Serial.write(controllerId);
  Serial.write((uint8_t*) &subscriptionDelay, sizeof(uint32_t));
  Serial.write(checksum);
}

void SubscriptionResponse::send() {
  Serial.write(messageId);
  Serial.write(controllerId);
  Serial.write(checksum);
}

void SubscriptionSensorUpdate::send() {
  Serial.write(messageId);
  Serial.write(controllerId);

  Serial.write(sensorTypeId);
  Serial.write((uint8_t*) &sensorReadingLength, sizeof(uint16_t));
  Serial.write(dataPtr, (size_t) sensorReadingLength);

  Serial.write(checksum);
}

void Error::send() {
  Serial.write(messageId);
  Serial.write(controllerId);
  Serial.write(errorCode);
  Serial.write(checksum);
}

//// MESSAGE RECEIVING /////////////////////////////////////////////////////////
//
// Receives an incoming hibike message.
// Returns a null pointer if no data is on the serial port.
//
// Note that Arduino's Serial.readBytes function automatically
// waits until either the given number of bytes are successfully
// read or a timeout occurs (defaults to 1000ms, but this is an 
// eternity for our use case so we should probably change this to a 
// few ms at most).
//
std::unique_ptr<HibikeMessage> receiveHibikeMessage() {
  if (!Serial.available()) {
    return nullptr;
  }
  //TODO: implement better error checking to detect/deal with possible failures
  uint8_t messageId, controllerId, checksum;
  HibikeMessage m;
  Serial.readBytes(&messageId, 1);
  Serial.readBytes(&controllerId, 1);
  switch (messageId) {
    case HibikeMessageType.SubscriptionRequest:
      uint32_t subscriptionDelay;
      Serial.readBytes(&subscriptionDelay, 4);
      m = new SubscriptionRequest(controllerId, subscriptionDelay);
      break;
    case Error:
      // TODO: add proper error handling dependant on errorCode
      uint8_t errorCode;
      Serial.readBytes(&errorCode, 1);
      m = new Error(controllerId, errorCode);
      break;
    default:
      // TODO: implement missing message types
      Error(controllerId, ErrorCode.InvalidMessageType).send();
  }
  Serial.readBytes(&checksum, 1);
  if (checksum ^ m.getChecksum()) {
    // send an error back to the main controller if the checksums aren't identical
    Error(controllerId, ErrorCode.ChecksumMismatch).send();
  }
  return unique_ptr<HibikeMessage>(m);
}

File: lib/hibike_message.py
import struct
from enum import Enum

"""
Message IDs for each type of hibike message.
"""
class HibikeMessageType(Enum):
    SubscriptionRequest  = 0x00
    SubscriptionResponse = 0x01
    SensorUpdate         = 0x02
    Error                = 0xFF

"""
Sensor Type IDs.
"""
class SensorType(Enum):
    LimitSwitch  = 0x00
    LineFollower = 0x01

"""
More specific error codes.
"""
class ErrorCode(Enum):
    InvalidMessageType = 0xFB
    MalformedMessage   = 0xFC
    InvalidArduinoId   = 0xFD
    ChecksumMismatch   = 0xFE
    GenericError       = 0xFF

class HibikeMessageException(Exception):
    def __init__(self, value):
        self.value = value
    def __str__(self):
        return repr(self.value)

def getByte(value, index):
    return (value >> (index*8)) & 0xFF


class HibikeMessage:
    def __init__(self, messageId, controllerId, serial = None):
        assert isinstance(messageId, HibikeMessageType)
        self.__messageId = messageId
        self.__controllerId = controllerId
        self.__checksum = None
        self.__serial = serial
    def getMessageId():
        return self.__messageId
    def getControllerId():
        return self.__controllerId
    def getChecksum():
        self.__calculateChecksum()
        return self.__checksum
    # relatively ghetto way of forcing overrides of __calculateChecksum() and send()
    # without using python's builtin abstract base class which has all sorts of weirdness
    def __calculateChecksum():
        raise HibikeMessageException("Must override the HibikeMessage __calculateChecksum method.");
    def send():
        raise HibikeMessageException("Must override the HibikeMessage send method.");

class SubscriptionRequest(HibikeMessage):
    def __init__(self, controllerId, subscriptionDelay, serial = None):
        HibikeMessage.__init__(self, HibikeMessageType.SubscriptionRequest,
                               controllerId, serial)
        # assert subscriptionDelay is a uint32
        self.__subscriptionDelay = subscriptionDelay
    def getSubscriptionDelay():
        return self.__subscriptionDelay
    def __calculateChecksum():
        if self.__checksum is not None:
            return
        self.__checksum = self.__messageId.value
        self.__checksum ^= self.__controllerId
        self.__checksum ^= getByte(self.__subscriptionDelay, 0)
        self.__checksum ^= getByte(self.__subscriptionDelay, 1)
        self.__checksum ^= getByte(self.__subscriptionDelay, 2)
        self.__checksum ^= getByte(self.__subscriptionDelay, 3)
    def send():
        assert serial is not None
        self.__calculateChecksum()
        self.__serial.write(struct.pack('<B', self.__messageId.value))
        self.__serial.write(struct.pack('<B', self.__controllerId))
        self.__serial.write(struct.pack('<I', self.__subscriptionDelay))
        self.__serial.write(struct.pack('<B', self.__checksum))


class SubscriptionResponse(HibikeMessage):
    def __init__(self, controllerId, serial = None):
        HibikeMessage.__init__(self, HibikeMessageType.SubscriptionResponse,
                               controllerId, serial)
    def __calculateChecksum():
        if self.__checksum is not None:
            return
        self.__checksum = self.__messageId.value
        self.__checksum ^= self.__controllerId
    def send():
        assert serial is not None
        self.__calculateChecksum()
        self.__serial.write(struct.pack('<B', self.__messageId.value))
        self.__serial.write(struct.pack('<B', self.__controllerId))
        self.__serial.write(struct.pack('<B', self.__checksum))

class SensorUpdate(HibikeMessage):
    def __init__(self, controllerId, sensorTypeId, sensorReadingLength, data, serial = None):
        HibikeMessage.__init__(self, HibikeMessageType.SensorUpdate, controllerId, serial)
        assert isinstance(sensorTypeId, SensorType)
        # assert uint8
        self.__sensorTypeId = sensorTypeId
        # assert uint16
        self.__sensorReadingLength = sensorReadingLength
        # assert size in bytes is consistent with sensorReadingLength
        self.__data = data
    def getSensorTypeId():
        return self.__sensorTypeId
    def getSensorReadingLength():
        return self.__sensorReadingLength
    def getData():
        return self.__data
    def __calculateChecksum():
        if self.__checksum is not None:
            return
        self.__checksum = self.__messageId.value
        self.__checksum ^= self.__controllerId
        self.__checksum ^= self.__sensorTypeId.value
        self.__checksum ^= getByte(self.__sensorReadingLength, 0)
        self.__checksum ^= getByte(self.__sensorReadingLength, 1)
        for i in range(self.__sensorReadingLength):
            self.__checksum ^= getByte(self.__data, i)
    def send():
        assert serial is not None
        self.__calculateChecksum()
        self.__serial.write(struct.pack('<B', self.__messageId.value))
        self.__serial.write(struct.pack('<B', self.__controllerId))
        self.__serial.write(struct.pack('<B', self.__sensorTypeId.value))
        self.__serial.write(struct.pack('<H', self.__sensorReadingLength))
        for i in range(self.__sensorReadingLength):
            self.__serial.write(struct.pack('<B', getByte(self.__data, i)))
        self.__serial.write(struct.pack('<B', self.__checksum))

class Error(HibikeMessage):
    def __init__(self, controllerId, errorCode, serial = None):
        HibikeMessage.__init__(self, HibikeMessageType.Error, controllerId, serial)
        assert isinstance(errorCode, ErrorCode)
        self.__errorCode = errorCode
    def getErrorCode():
        return self.__errorCode
    def __calculateChecksum():
        if self.__checksum is not None:
            return
        self.__checksum = self.__messageId.value
        self.__checksum ^= self.__controllerId
        self.__checksum ^= self.__errorCode.value
    def send():
        assert serial is not None
        self.__calculateChecksum()
        self.__serial.write(struct.pack('<B', self.__messageId.value))
        self.__serial.write(struct.pack('<B', self.__controllerId))
        self.__serial.write(struct.pack('<B', self.__errorCode.value))
        self.__serial.write(struct.pack('<B', self.__checksum))

def receiveHibikeMessage(serial):
    assert serial is not None
    if serial.inWaiting() is 0:
        return None
    m = None
    messageId = HibikeMessageType(struct.unpack('<B', serial.read(1))[0])
    controllerId = struct.unpack('<B', serial.read(1))

    if messageId is HibikeMessageType.SubscriptionResponse:
        m = SubscriptionResponse(controllerId, serial)

    elif messageId is HibikeMessageType.SensorUpdate:
        sensorTypeId = SensorType(struct.unpack('<B', serial.read(1))[0])
        sensorReadingLength = struct.unpack('<H', serial.read(2))
        data = 0
        for i in range(sensorReadingLength):
            #FIXME: check on endianness
            data = data << 8 | struct.unpack('<B', serial.read(1))
        m = SensorUpdate(controllerId, sensorTypeId, sensorReadingLength, data, serial)

    elif messageId is HibikeMessageType.Error:
        errorCode = ErrorCode(struct.unpack('<B', serial.read(1))[0])
        m = Error(controllerId, errorCode, serial)

    checksum = struct.unpack('<B', serial.read(1))
    if checksum ^ m.getChecksum() is not 0:
        Error(controllerId, ErrorCode.ChecksumMismatch, serial).send()
        return None
    return m
